### PR TITLE
Guide on "Don't block the event loop"

### PR DESCRIPTION
**Pull request by davisjam**
Towards #1471.

Not finished yet, hoping to get some initial feedback.

### CODE REVIEW

**Review by Tiriel — COMMENTED** [review]
Haven't finished ready in entirety yet, but one general opinion yet: I think you should go for a more formal tone.
This sounds too much like a blog post to me, not like an "official" guide. That may not be the general feeling of everyone, but I for one would prefer a more neutral tone on a guide found on Node.js official website.

Coming soon: complete review, although I'm not as technical as some!

**Comment by davisjam** [comment]
Thanks for the comment on tone. I will try a more formal approach in my next revision.

Thoughts about the technical content?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (Tiriel) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: Quick question here, mostly for my own education: 
In that case, wouldn't `stringVar.indexOf(searchVar) !== -1` or `stringVar.includes(searchVar) !== false` be more efficient, and eliminating the need of another module?

I originally come from PHP, and replacing `preg_match()` with `stripos($string, $search) !== false` is mostly preferred for the same reasons, so I thought it would be the same in JS.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (Tiriel) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: Some example might be good here. I think most newcomers/basic users won't necessarily have a clue about what you're talking here. The notion of 'immediate' in the event loop can be misleading and some people won't understand why it would partly solve the problem IMHO.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (Tiriel) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: Damn, that was the piece I was most eager to read :smile:

**Review by Tiriel — COMMENTED** [review]
As said, I'm not the most technical reviewer you could get, I'm largely a noob in Node.js and JS itself. I still hope my comments will help!

**Comment by vsemozhetbyt** [comment]
@davisjam It seems your local email in git settings does not match your GitHub one. Can you add a local one to GitHub email list so that your contributions to Node.js repositories can be associated with your GitHub account?

mail

**Comment by davisjam** [comment]
@vsemozhetbyt Thanks, always forget this.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: If you're just doing a `strcmp` or `substr`-style string match, then yes, you should just use `indexOf`. You don't need a regular expression for this, nor will a regexp actually be dangerous in this context.

But sometimes people need more matching power for more complicated expressions, and a sufficiently complicated regexp can lead to REDOS.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Good idea. I've been thinking about some examples.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Soon!

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (Tiriel) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Okay, that's reassuring :smile: 
Maybe in that case you could specify this alternative, and that sufficiently complex regex are the real problem here? (I'm not quite sure what a sufficiently complete regex would be either; regex using escaped/reserved chars and/or logn regex?)

Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Is the new REDOS section clearer? Still needs work?

**Comment by davisjam** [comment]
@Tiriel:
1. Is the language up to an appropriate level of formality, or should I work on the tone further?
2. Does the regexp section make more sense now?

I haven't revised anything after the REDOS section yet.

**Comment by Tiriel** [comment]
@davisjam Way better now IMHO! It's more precise, the examples add a lot, and the tone is better suited to the subject and target as I see it.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Alrighty, first draft of this is up.

**Comment by davisjam** [comment]
The first draft of the full document is up. Comments appreciated.

**Comment by ayaankazerouni** [comment]
+1 Cool guide! I learned a lot, especially from the REDOS section.

**Comment by davisjam** [comment]
> I'm not the most technical reviewer you could get, I'm largely a noob in Node.js and JS itself

@Tiriel When you have a moment, can you ping some appropriate folks for technical feedback? Your thoughts on style and correctness are of course also much appreciated.

**Comment by vsemozhetbyt** [comment]
Let's cc @nodejs/documentation for a start.

**Comment by gibfahn** [comment]
cc/ @sam-github if you're around, seems like something you'd be interested in.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: cli scripts can do, for example, heavy DNS lookups at the same time as file I/O, so worker pool can be blocked.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: No, just file I/O, network I/O is handled on the main loop.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: where small number of threads is likely zero thread for a network server (no file i/o

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: file i/o isn't expensive, no more so than network i/o, it just doesn't have good completion APIs at the O/S level to allow it to be done on the main loop, so threads are needed so that i/o can be done with blocking system calls

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: the o/s produces the events, not the node runtime

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: only on linux, its kqueue on OS X, and other mechanisms on Windows

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: this isn't the common understanding of cooperative multi-tasking. coop tasking means that the running thread can suspend itself, but node doesn't do this, ever (on main loop or in pool). it runs to completion, which is not cooperative multitasking

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: what about when you have multiple addons, each with a number of threads equal to the number of logical cores?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: only dns.lookup, the other ones don't use the pool

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: the thread in the libuv work pool aren't called Event Handlers, except here, by you! And the lib uv workers never execute JS code. I don't think its helpful to overgeneralize the main thread to describe it as a kind of worker.

**Review by sam-github — COMMENTED** [review]
Its good to provide some information on this, particularly to provide implementation advice to users. It would be easy for someone to read this and think that network I/O occurs in the uv thread pool (which it doesn't), I think that aspect should be much more clear.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: True. I view the server case as more concerning because of the DoS possibility. Do you want me to remove this caveat?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Yes. I'll fix to clarify what kinds of I/O go to the Worker Pool.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: This depends on the env variable and on whether the server has ever used the threadpool, since libuv's threads are created the first time the threadpool is used. I don't see a reason to confuse the issue though, this is intended as the high-level summary.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Remove, or expand to make more clear the situations in which some things aren't problems. No need to optimize anything for a short-running app that does a one-shot HTTP request is what you are trying to say, I think.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: I feel this confuses the issue, though, because the secret to node's scalability isn't that it uses threadpools well (so does java), the secret is that a fairly typical HTTP server making outbound DB or other network requests uses *no threads at all*.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: My understanding of the libuv loop (which is limited to Linux) is that the trouble is not the completion APIs but rather the ability of the OS to say that the file descriptor is ready and thus not to block when you do I/O on it. libuv only sends network traffic when it will not block in doing so, but the OS will not guarantee that read/write on a file descriptor will be non-blocking.

Based on this understanding, my claim is correct - the I/O done on the event loop is always cheap *from the event loop's perspective* because the OS guarantees it to be non-blocking.

Am I missing something here?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Yes, via things like epoll. Where do you see the line between technical accuracy and "missing the forest for the trees" in guides like this?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Yes, good catch.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: I couldn't think of a better way to describe this other than saying "queue-based". Should I stick with that? Is there a better term?

Counterpoint, though -- queue-based is "cooperative multitasking" in the sense that a callback/task yields precisely when it completes.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: I would suggest creating a single Computation Worker Pool at the application level, similar to the Node Worker Pool that handles I/O from all points in the application. Do you have any reference applications I can look at to describe alternate patterns?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Maybe. The actual problem that is that you can't poll/epoll/select/kqueue a file descriptor for readiness, and so we can't do non-blocking reads or writes. And I guess you could call a blocking call "expensive", but network i/o can be done in thread pools with blocking calls, and lots of languages do exactly that. node doesn't, because network i/o can be non-blocking on unix, whereas file i/o can't (without the async i/o posix extensions, but they are poorly supported, and while windows has better async i/o for files, early tests by libuv authors showed that it would randomly block, it was to flaky to be useful).

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Yes. I'll fix this. I believe all of the FS APIs map directly to libuv versions, though?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Yes, I agree that this is a weird overgeneralization. I'll take another look at this.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: What do you think about saying "an epoll-like system call" with a link to the epoll man page? Part of my goal is to point users to resources that will give them a richer understanding of Node concepts and Node under the hood. That was also my intent with the "cooperative multitasking" link.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: because its not really a queue, if it was, we could ask "how long is that queue?", but we can't.... because it doesn't exist. node reads from the o/s's internal queue, gets a set of ready events, processes them then blocks reading again. no queue. the whole queue thing is brought up sometimes, but I don't think introducing something that doesn' exist is helpful in understanding how node works. I never describe anything like it in https://www.youtube.com/watch?v=P9csgxBgaZ8, for example

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Right. It uses *one* thread -- the main loop -- to handle many clients, rather than Apache-style with one thread per active client.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: but no such thing exists, so current addons don't use it. it has been proposed that uv expose two work APIs, one for blocking non-cpu intensive calls (file i/o, dns.lookup), the other for cpu intensive activities (crypto.pbkdf for key generation, etc), but no code has been PRed, or benchmarks to show it will work. so, in the meantime, you have to consider all the internal pools of all the addons, and actually most addons will mix cpu and i/o intensive work in the same uv work pool. I'm slowly working towards getting metrics in node about the state of the uv pool, but that doesn't exist yet, so not worth mentioning here.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Last I checked, posix async I/O is just a threadpool in libc, no different from the libuv threadpool. I've prototyped Node with kernel AIO to see if things had improved since [2011](https://github.com/nodejs/node-v0.x-archive/issues/1446), but the O_DIRECT restriction remains and its costs are catastrophic.

Anyway, I'll flesh out this discussion a bit more.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: https://nodejs.org/api/fs.html#fs_threadpool_usage  - except fs.FSWatcher()

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: OK. If I added such an API would it be of interest? Is anyone working on this right now?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: aio was intended to be implemented in kernel and thus to rock... but hasn't happened yet, the saga is long, some info here: https://lwn.net/Articles/724198/. There is various user-space emulation using blocking I/O in a thread pool....  so, yeah, uv doesn't use that, there is no point.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Also, I'll just remove this section. It sounds like this is too much of an open problem to worry about right now.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: I think that other than the queue to the uv work pool (which is actually a queue), I don't think queues should be mentioned. Not sure what a good replacement is, sorry! I can see the ghostly parallels, both node and co-op multitasking systems require careful consideration of how much work you do, and how you construct your app, but the analogy breaks down fast, because cooperative multi-tasking has a fairly agreed upon definition, and its not what node does, so better stay away from it, IMO.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: From a framework developer perspective, it's a queue that you can only see looking backwards. From an application developer perspective, though, I think the concept of a queue is a fine approximation of reality. I'm envisioning the target audience to be application developers for whom I think the "queue" concept is fine.

How strongly do you feel about this?

**Comment by davisjam** [comment]
Thanks for all the great comments, @sam-github! I'll get started on a revision.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: I am OK with the whole thing, except this sentence:

>The Node runtime produces events for this queue.

I think that specific sentence should be removed. node doesn't produce the events. it asks for the o/s to produce them. I wouldn't say that I'm producing a cappucino when I order a cappucino at a cafe. Causing it to be produced? Sure, producing it? no.

**Comment by Tiriel** [comment]
Just a quick semi off-topic:
Thanks @sam-github & @davisjam , just reading your comments makes me learn a huge load of new things!
I really look forward to reading the final version of this guide!

**Comment by davisjam** [comment]
@sam-github e9c40e9 attempts to address your concerns.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (a0viedo) [inline locale/en/docs/guides/dont-block-the-event-loop.md:58]: worth mentioning `net` module?

Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md:58]: It is my understanding that the [`net`](https://nodejs.org/api/net.html) module does *not* use the Worker Pool, thanks to good OS support for non-blocking network I/O. The module docs do not bear the usual caveats about the threadpool, and as far as I can tell from the source there is no call to `uv_queue_work`.

Am I wrong in this? Or are you suggesting I explicitly mention the `net` module as a "safe" option? If so, I attempted to do so in the section "What code runs on the Event Loop?" with the sentence "The Event Loop will also fulfill the non-blocking asynchronous requests made by its callbacks, e.g., network I/O.". Do you think I should name a few examples there, like `net` and `UDP`?

Comment (sam-github) [reply locale/en/docs/guides/dont-block-the-event-loop.md:58]: correct, net does not use threads

**Comment by sam-github** [comment]
thanks for addressing my comments @davisjam

**Comment by davisjam** [comment]
Sure thing, @sam-github. Any more suggestions, or people that should be tagged?

**Comment by sam-github** [comment]
@jasnell was looking at loop behaviour, he might have some comments

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (benjamingr) [inline locale/en/docs/guides/dont-block-the-event-loop.md:470]: What about the JavaScript thread?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (benjamingr) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: This is actually being fixed if I understand correctly: https://github.com/nodejs/node/pull/17054

Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: That's the idea, anyway. But this guide is describing things as they are, not as I hope they will be.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md:470]: I am referring to the "JavaScript thread" as the Event Loop throughout, to contrast it with the Worker Pool (threadpool) threads, and because it does more than JavaScript (e.g. wandering into libuv to fulfill network I/O requests, timers, etc.).

See the section "What code runs on the Event Loop?"

"In summary, the Event Loop executes the JavaScript callbacks registered for events, and is also responsible for fulfilling non-blocking asynchronous requests like network I/O."

Let me know if this is not clear or if you have alternative suggestions for phrasing.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: See also nodejs/node#17250.

Comment (gibfahn) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Might be worth adding a link to nodejs/node#17054 in here to note the proposed change then.

Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Added a link in 0a42816, thank you.

**Comment by davisjam** [comment]
I recognize this is a pretty long document, but does anyone have additional change requests?

**Review by benjamingr — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (gibfahn) [inline locale/en/docs/guides/dont-block-the-event-loop.md:250]: Is it worth mentioning asynchronous JSON parsing modules here (and then linking to `The risks of npm modules` section)?

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (gibfahn) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: What do you do if there is no async core api?

Maybe the answer is "raise an issue in core", but this was my question on reading this, so maybe worth saying something.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (gibfahn) [inline locale/en/docs/guides/dont-block-the-event-loop.md]: Does node core undergo a formal verification process? Mentioning it here makes it seem like it's in contrast to something else, but I'm not sure what that something is.

**Review by gibfahn — COMMENTED** [review]
This is really helpful, thanks a lot for writing this!

LGTM, couple of questions though.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md:250]: Good idea, will do.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: You're right, "formal" isn't the right word. It's more of a "many eyes on Node, few eyes on most 3rd-party modules" type situation. I'll change the wording.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Thanks, I'll try another pass on this.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md:250]: Addressed in ccd4f75.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Addressed in ccd4f75.

File: locale/en/docs/guides/dont-block-the-event-loop.md
Comment (davisjam) [reply locale/en/docs/guides/dont-block-the-event-loop.md]: Addressed in ccd4f75.

**Comment by davisjam** [comment]
@gibfahn Thank you for your review. I've tried to address your comments in ccd4f75. Let me know if you have more concerns before you'd be willing to approve.

**Comment by davisjam** [comment]
@sam-github Anything else needed for your approval?

**Comment by sam-github** [comment]
looks good to me

**Review by gibfahn — APPROVED** [review]

**Comment by davisjam** [comment]
The full draft has been up for about 2 weeks and has gone through a few reviews. Ready to be landed?

**Pull request merged**
